docs/architecture/adr-005-modular-server.md

```diff
@@ -0,0 +1,169 @@
+# ADR 005: Modular Server
+
+## Changelog
+
+- 2020/12/22: Initial Draft
+
+## Author
+
+- Federico Kunze ([@fedekunze](https://github.com/fedekunze))
+
+## Status
+
+Draft, Not Implemented
+
+## Abstract
+
+This document outlines a modular and standard approach for using servers on the SDK, so that
+applications can also run custom services besides the default ones provided by the SDK.
+
+## Context
+
+The current `Server` implementation allows for applications to register routes, API and gRPC
+services to the SDK. The problem with the current approach is that the SDK uses a concrete type,
+which prevents other SDK-based blockchains to extend the services provided by the application
+without forking the application `start` command (eg: `<app>d start`).
+
+This modularity and extensibility relies on the following limitations:
+
+    - Consistent services
+    - App configuration extensibility
+
+### Consistent services
+
+The `start` command executes 3 steps for each of the in-process services provided by an SDK
+application server:
+
+    1. Creation of the service
+    2. Registration of the services directly with the server
+    3. Start/Stop the service process
+  
+In the current implementation, [some services](https://github.com/cosmos/cosmos-sdk/blob/f9dc082059d63423f96315ff913a8745c3178a7d/server/grpc/server.go#L14-L42) wrap these steps together in a single function, while
+[others](https://github.com/cosmos/cosmos-sdk/blob/f9dc082059d63423f96315ff913a8745c3178a7d/server/api/server.go#L81-L119) implement them as separate functions as the ones above. The lack of a standard approach for
+these steps results in difficulty for extensibility as each of these functions are individually
+called by the `start` command after checking if the service is enabled or not by the configuration.
+
+### App configuration extensibility
+
+Each service relies on the configuration options defined on `app.toml`. Thus, in order to extend
+the services provided by the server, the SDK must handle custom TOML files provided by each
+application.
```

Comment:
If we fork a service, then we don't need to create a new config file. Same, if we are able to extend Server / service in a way to configure which services are registered. So, we can keep all configuration in `app.toml`. The problem is to manage properly which services are registered in server, right?